File: leetcode/23-merge-k-sorted-lists/solution.java
import java.util.Comparator;
import java.util.PriorityQueue;

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        ListNode node = null;
		final PriorityQueue<Integer> q = new PriorityQueue<>(Comparator.reverseOrder());

		for (ListNode l : lists) {
            node = l;
			while (node != null) {
				q.add(node.val);
				node = node.next;
			}
		}

	    var val = q.poll(); 	

        while (val != null) {
            node = new ListNode(val, node);
            val = q.poll();
        }
		
        return node;
        
    }
}

File: leetcode/23-merge-k-sorted-lists/solution.cpp
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
      ListNode* node = NULL;
      priority_queue<int> q;

      for (int i = 0; i<lists.size();i++) {
        node = lists[i];

        while (node != NULL) {
          q.push(node->val);
          node = node->next;
        }
      }

      while (!q.empty()) {
        int val = q.top();
        q.pop();
        node = new ListNode(val, node);
      }

      return node;
        
    }
};
